### PR TITLE
[serializers] Allow setting request/session later

### PR DESCRIPTION
**Pull request by LiraNuna**
`request` and `session` may not be available at serialization time or in effort to reduce boiler plate it would be wise to set them later on the request pipeline.

This PR allows the `request` and `session` attribute to be assigned later in order to achieve this.

### CODE REVIEW

File: cartographer/serializers/schema_serializer.py [hunk 1]
```diff
@@ -45,36 +45,62 @@ def __init__(self, model,
...
+
+        self._session = session
+
+    @property
+    def includes(self):
+        return self._includes
+
+    @includes.setter
+    def includes(self, includes):
+        if not includes:
```
Comment (21echoes) [inline cartographer/serializers/schema_serializer.py]: the empty array is falsy, and we don't want to turn it into `default_includes`

File: cartographer/serializers/schema_serializer.py [hunk 2]
```diff
@@ -45,36 +45,62 @@ def __init__(self, model,
...
+
         self._linked_resources = None
         self._masked_fields = None
         self._masked_includes = None
 
         self.prime_for_includes()
 
+    @property
+    def request(self):
+        return self._request
```
Comment (21echoes) [inline cartographer/serializers/schema_serializer.py]: previously, we never held on to the request. I don't think we should start, as once the query params are pulled off of the request, we don't need it any more.

File: cartographer/serializers/schema_serializer.py [hunk 3]
```diff
@@ -45,36 +45,62 @@ def __init__(self, model,
...
+            self.includes = request.get_includes()
+
+        if not self.requested_fields and request:
+            self.requested_fields = request.get_requested_fields()
+
+        self._request = request
+
+    @property
+    def session(self):
+        return self._session
```
Comment (21echoes) [inline cartographer/serializers/schema_serializer.py]: as with the above: previously, we never held on to the session. I don't think we should start, as once the current user id is pulled off the session, we don't need it any more.

File: cartographer/serializers/schema_serializer.py [hunk 4]
```diff
@@ -45,36 +45,62 @@ def __init__(self, model,
...
+        if includes is None and parent_serializer is not None and relationship_name is not None:
+            prefix = relationship_name + '.'
+            includes = [include_path[len(prefix):]
+                        for include_path in parent_serializer.includes
+                        if include_path.startswith(prefix)]
 
-        if requested_fields is None and inbound_request:
-            requested_fields = inbound_request.get_requested_fields()
+        self.includes = includes
+        self.parent_serializer = parent_serializer
```
Comment (21echoes) [inline cartographer/serializers/schema_serializer.py]: holding on to the `parent_serializer` seems strange to me

**Comment by 21echoes** [comment]
`inbound_request` and `inbound_session` were supposed to be helpers, not properties. I'd like to move this to something more like `serializer.parse_query_params_from_request(request)` and `serializer.parse_session(session)`

**Comment by alts** [comment]
What's the motivation broader motivation for this? Why is deferring assignment of request/session a good/needed thing?

**Comment by 21echoes** [comment]
also curious about @alts 's question -- most `Serializer` objects are not particularly long lived (in fact, most of them live for the duration of literally one line of code). This is more or less by design -- they can't do much other than get serialized. If we have a really good reason to do this, I'm open to it, but I'd be decently surprised if we needed to pass around these objects, much less build them up across many different functions.

**Comment by LiraNuna** [comment]
You raise good points... I can try taking a different approach

**Pull request closed without merging**